Mathlib/LinearAlgebra/TensorProduct/RightExactness.lean

```diff
@@ -0,0 +1,648 @@
+  rw [LinearMap.range_le_iff_comap, ← LinearMap.ker_comp, ← rTensor_comp,
+    hfg.linearMap_comp_eq_zero, rTensor_zero, ker_zero]
+
+/-- The inverse map in `rTensor.equiv_ofRightInverse`
+  (computably, given a right inverse) -/
+def rTensor.inverse_ofRightInverse {h : P → N} (hgh : Function.RightInverse h g) :
+    P ⊗[R] Q →ₗ[R] N ⊗[R] Q ⧸ LinearMap.range (rTensor Q f) := by
+  rw [exact_iff] at hfg
+  refine
+    TensorProduct.lift
+      { toFun := fun p ↦ Submodule.mkQ _ ∘ₗ TensorProduct.mk R _ _ (h p)
+        map_add' := fun p p' => LinearMap.ext <| fun q => (Submodule.Quotient.eq _).mpr ?_
+        map_smul' := fun r p => LinearMap.ext <| fun q => (Submodule.Quotient.eq _).mpr ?_ }
+  · change h (p + p') ⊗ₜ[R] q - (h p ⊗ₜ[R] q + h p' ⊗ₜ[R] q) ∈ range (rTensor Q f)
+    rw [← TensorProduct.add_tmul, ← TensorProduct.sub_tmul]
+    apply le_comap_range_rTensor f
+    rw [← hfg, mem_ker, map_sub, map_add]
+    simp only [hgh _, sub_self]
+  · change h (r • p) ⊗ₜ[R] q - r • h p ⊗ₜ[R] q ∈ range (rTensor Q f)
+    rw [TensorProduct.smul_tmul', ← TensorProduct.sub_tmul]
+    apply le_comap_range_rTensor f
+    rw [← hfg, mem_ker, map_sub, map_smul]
+    simp only [hgh _, sub_self]
+
+lemma rTensor.inverse_ofRightInverse_apply
+    {h : P → N} (hgh : Function.RightInverse h g) (y : N ⊗[R] Q) :
+    (rTensor.inverse_ofRightInverse Q hfg hgh) ((rTensor Q g) y) =
+      Submodule.Quotient.mk (p := LinearMap.range (rTensor Q f)) y := by
+  simp only [← LinearMap.comp_apply, ← Submodule.mkQ_apply]
+  rw [exact_iff] at hfg
+  apply LinearMap.congr_fun
+  apply TensorProduct.ext'
+  intro n q
+  simp only [coe_comp, Function.comp_apply, rTensor_tmul, Submodule.mkQ_apply]
+  simp only [rTensor.inverse_ofRightInverse]
+  simp only [TensorProduct.lift.tmul, coe_mk, AddHom.coe_mk, mk₂_apply]
+  simp only [coe_comp, Function.comp_apply, mk_apply, Submodule.mkQ_apply]
+  rw [Submodule.Quotient.eq]
+  rw [← TensorProduct.sub_tmul]
+  apply le_comap_range_rTensor f
+  rw [← hfg, mem_ker, map_sub, sub_eq_zero, hgh]
+
+/-- The inverse map in `rTensor.equiv` -/
+noncomputable
+def rTensor.inverse :
+    P ⊗[R] Q →ₗ[R] N ⊗[R] Q ⧸ LinearMap.range (rTensor Q f) :=
+  rTensor.inverse_ofRightInverse Q hfg (Function.rightInverse_surjInv hg)
+
+lemma rTensor.inverse_apply (y : N ⊗[R] Q) :
+    (rTensor.inverse Q hfg hg) ((rTensor Q g) y) =
+      Submodule.Quotient.mk (p := LinearMap.range (rTensor Q f)) y := by
+  rw [rTensor.inverse, rTensor.inverse_ofRightInverse_apply]
+
+lemma rTensor.inverse_comp_rTensor :
+    (rTensor.inverse Q hfg hg).comp (rTensor Q g) =
+      Submodule.mkQ (p := LinearMap.range (rTensor Q f)) := by
+  rw [LinearMap.ext_iff]
+  intro y
+  simp only [coe_comp, Function.comp_apply, Submodule.mkQ_apply, rTensor.inverse_apply]
+
+/-- For a surjective `f : N →ₗ[R] P`,
+  the natural equivalence between `N ⊗[R] Q ⧸ (image of ker f)` to `P ⊗[R] Q`
+  (computably, given a left inverse) -/
+def rTensor.equiv_ofRightInverse
```

Comment:
I never remember which is the left inverse and which is the right inverse but given that you've called this function `...ofRightInverse` I'm surprised you're talking about left inverse in the docstring.

I think it should be called `def rTensor.linearEquiv_of_rightInverse` though?

Reply:
Adjusted